File: python/open_guardrail/guards/response_structure_json.py
"""Validate expected JSON structure in responses."""
from __future__ import annotations

import json
import time
from typing import List, Optional

from open_guardrail.core import GuardResult


class _ResponseStructureJson:
    def __init__(
        self,
        *,
        action: str = "block",
        required_keys: Optional[List[str]] = None,
    ) -> None:
        self.name = "response-structure-json"
        self.action = action
        self._required = required_keys or []

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        try:
            parsed = json.loads(text.strip())
        except (json.JSONDecodeError, ValueError):
            issues.append("invalid-json")
            elapsed = (
                time.perf_counter() - start
            ) * 1000
            return GuardResult(
                guard_name=(
                    "response-structure-json"
                ),
                passed=False,
                action=self.action,
                score=1.0,
                message="Invalid JSON",
                latency_ms=round(elapsed, 2),
                details={"issues": issues},
            )

        if not isinstance(parsed, dict):
            issues.append("not-object")
        else:
            for key in self._required:
                if key not in parsed:
                    issues.append(
                        f"missing-key:{key}"
                    )
                elif parsed[key] is None:
                    issues.append(
                        f"null-value:{key}"
                    )

        triggered = len(issues) > 0
        denom = len(self._required) or 1
        score = (
            min(len(issues) / denom, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (
            time.perf_counter() - start
        ) * 1000

        return GuardResult(
            guard_name="response-structure-json",
            passed=not triggered,
            action=(
                self.action if triggered else "allow"
            ),
            score=score,
            message=(
                "JSON structure invalid"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues}
                if triggered
                else None
            ),
        )


def response_structure_json(
    *,
    action: str = "block",
    required_keys: Optional[List[str]] = None,
) -> _ResponseStructureJson:
    return _ResponseStructureJson(
        action=action,
        required_keys=required_keys,
    )

File: packages/guards/src/response-structure-json.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface ResponseStructureJsonOptions {
  action: 'block' | 'warn';
  requiredKeys?: string[];
}

export function responseStructureJson(options: ResponseStructureJsonOptions): Guard {
  const requiredKeys = options.requiredKeys ?? [];

  return {
    name: 'response-structure-json',
    version: '0.1.0',
    description: 'Validates expected JSON structure in responses',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      let parsed: unknown;
      try {
        parsed = JSON.parse(text.trim());
      } catch {
        issues.push('invalid-json');
        return {
          guardName: 'response-structure-json',
          passed: false,
          action: options.action,
          score: 1.0,
          latencyMs: Math.round(performance.now() - start),
          details: { issues },
        };
      }

      if (typeof parsed !== 'object' || parsed === null || Array.isArray(parsed)) {
        issues.push('not-object');
      } else {
        const obj = parsed as Record<string, unknown>;
        for (const key of requiredKeys) {
          if (!(key in obj)) {
            issues.push(`missing-key:${key}`);
          } else if (obj[key] === null || obj[key] === undefined) {
            issues.push(`null-value:${key}`);
          }
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / requiredKeys.length || 1, 1.0) : 0;

      return {
        guardName: 'response-structure-json',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues } : undefined,
      };
    },
  };
}
